### PR TITLE
Add Searchfunction / SearchPage

### LINKED ISSUE #37: Add SearchResults-Page
# User story
As a **user**,
I want to have a SearchResults-Page
so I can see the tasting events I am looking for

# Description

# Notes / Links

# Acceptance Criteria
- [x] Searchbar in header
- [x] Back button
- [x] Cancel button
- [x] EventCard (Search result)
- [ ] Join Button (Part of search result)
- [x] Footer with respective Icons

### PR DESCRIPTION
**Pull request by rheimers**
Fix #37 

Done: 
- Submit Search on EventPage to SearchPage
- Display Results on SearchPage
- Add respective Routes to App.js
- Add canceling function to Searchbar 

Pending: 
- Fix Styling SearchPage
- Add Join Button to SearchResult

Bildschirmfoto 2020-10-08 um 07 41 30

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -1,14 +1,29 @@
 import React from "react";
-//import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
+import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
 import GlobalStyles from "./GlobalStyles";
 import EventPage from "./pages/EventPage";
+import SearchPage from "./pages/SearchPage";
+import styled from "@emotion/styled";
+
+const AppContainer = styled.div`
+  margin-bottom: 100px;
```
Comment (LebkuchenCJ) [inline client/src/App.js]: Can you adapt this to REM or %?
When I unclick it, there seems to be no effect: 
image

File: client/src/components/SearchPageHeader.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import React, { useState } from "react";
+import Searchbar from "../components/Searchbar";
+import styled from "@emotion/styled";
+import ArrowLeft from "../assets/icons/arrowleft.svg";
```
Comment (LebkuchenCJ) [inline client/src/components/SearchPageHeader.js]: For me it is always easier to understand code if you add a prefix to a variable as in to suggest what is behind that variable
```suggestion
import ArrowLeftSrc from "../assets/icons/arrowleft.svg";
```

File: client/src/components/Searchbar.js [hunk 3]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
   min-width: 284px;
   background-image: url(${SearchIcon});
   background-repeat: no-repeat;
-  background-position: 8px 7px;
+  background-position: 12px 10px;
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js]: See comment above regarding px value

File: client/src/components/Searchbar.js [hunk 4]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
   font: var(--font-robotolight);
 `;
 
-export default function Searchbar() {
+const ClearButton = styled.div`
+  position: absolute;
+  top: 57px;
+  right: 57px;
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js]: See comment above

File: client/src/components/Searchbar.js [hunk 5]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
+  background: transparent;
+  border: none;
+  &:focus {
+    outline: none;
+  }
+`;
+
+export default function Searchbar({ onSubmit, onClear, ...props }) {
+  const handleClear = (e) => {
+    e.preventDefault();
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js]: Here you `e` and above you use `event`. Try to be consistent with your code 💯

File: client/src/components/Searchbar.js [hunk 6]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
+    console.log("onClear", onClear);
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js:39]: Do you still need this console.log?

File: client/src/components/Searchbar.js [hunk 7]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
+    onClear();
+  };
   return (
-    <InputContainer>
-      <Input placeholder="Search for any tasting" />
-    </InputContainer>
+    <InputForm onSubmit={onSubmit}>
+      <Input placeholder="Search for any tasting" {...props} />
+      <ClearButton onClick={handleClear}>
+        <img src={CancelIcon} alt="Cancel" />
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js:46]: See comment from above. Here you are using "..Icon" and above only the variable

File: client/src/components/Searchbar.js [hunk 8]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
+  &:focus {
+    outline: none;
+  }
+`;
+
+export default function Searchbar({ onSubmit, onClear, ...props }) {
+  const handleClear = (e) => {
+    e.preventDefault();
+    console.log("onClear", onClear);
+    onClear();
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js:40]: OnClear seems to work but also there will be a function as output. Maybe you could check it again?
image

File: client/src/components/Searchbar.js [hunk 9]
```diff
@@ -15,14 +18,38 @@ const Input = styled.input`
...
+      <ClearButton onClick={handleClear}>
+        <img src={CancelIcon} alt="Cancel" />
+      </ClearButton>
+    </InputForm>
   );
 }
+
+Searchbar.propTypes = {
+  onSubmit: PropTypes.string.isRequired,
+  onClear: PropTypes.string.isRequired,
```
Comment (LebkuchenCJ) [inline client/src/components/Searchbar.js:54]: Be careful here.  Neither of these are a string. 
```suggestion
  onSubmit: PropTypes.func.isRequired,
  onClear: PropTypes.func.isRequired,
```
In general, onXXXX always suggest an event. A string could be named eg.: query, dataInput, value or something else

File: client/src/hooks/useAsync.js [hunk 10]
```diff
@@ -11,7 +11,7 @@ function useAsync(asyncFunction) {
         setLoading(true);
         setError(false);
         setData(null);
-        const data = await asyncFunction();
+        const data = await asyncFunction(args);
```
Comment (LebkuchenCJ) [inline client/src/hooks/useAsync.js:14]: 🥇

File: client/src/pages/EventPage.js [hunk 11]
```diff
@@ -14,11 +15,21 @@ const Container = styled.div`
...
+    history.push(`/search?q=${query}`);
+  }
+
   return (
     <Container>
-      <Searchbar></Searchbar>
+      <Searchbar
+        value={query}
+        onChange={(event) => setQuery(event.target.value)}
+        onSubmit={handleSubmit}
```
Comment (LebkuchenCJ) [inline client/src/pages/EventPage.js:31]: You could think of setting the state and defining the function in your component itself for a better re-use purpose

File: client/src/pages/SearchPage.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    return new URLSearchParams(useLocation().search);
+  };
+  let query = useQuery();
+  console.log(query);
+  const { data: events } = useAsync(getFilteredEvents, query.get("q"));
+  console.log("[SearchPage] events:", events);
+
+  return (
+    <Container>
+      <SearchPageHeader></SearchPageHeader>
```
Comment (LebkuchenCJ) [inline client/src/pages/SearchPage.js]:
```suggestion
      <SearchPageHeader/>
```

File: client/src/pages/SearchPage.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+`;
+
+function SearchPage() {
+  const useQuery = () => {
+    return new URLSearchParams(useLocation().search);
+  };
+  let query = useQuery();
+  console.log(query);
+  const { data: events } = useAsync(getFilteredEvents, query.get("q"));
+  console.log("[SearchPage] events:", events);
```
Comment (LebkuchenCJ) [inline client/src/pages/SearchPage.js]: Do you still need these console.logs?

File: client/src/pages/SearchPage.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  margin-left: 0.8em;
+`;
+
+function SearchPage() {
+  const useQuery = () => {
+    return new URLSearchParams(useLocation().search);
+  };
+  let query = useQuery();
+  console.log(query);
+  const { data: events } = useAsync(getFilteredEvents, query.get("q"));
```
Comment (LebkuchenCJ) [inline client/src/pages/SearchPage.js:21]: Why `.get("q")` ?

File: client/src/pages/SearchPage.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  console.log("[SearchPage] events:", events);
+
+  return (
+    <Container>
+      <SearchPageHeader></SearchPageHeader>
+      {events?.map((item, id) => (
```
Comment (LebkuchenCJ) [inline client/src/pages/SearchPage.js]: Could item be renamed in "event" maybe? As I understand you are importing events and mapping out a single event, or?

File: client/src/pages/SearchPage.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        <EventCard
+          category={item.category}
+          key={id}
+          title={item.title}
+          date={new Date(item.date)}
+          imgSrc={item.imgSrc}
+        />
+      ))}
+
+      <Footer></Footer>
```
Comment (LebkuchenCJ) [inline client/src/pages/SearchPage.js]:
```suggestion
      <Footer/>
```

**Review by LebkuchenCJ — COMMENTED** [review]
Hi 👍 Gj so far 💯 
I found a few things you should check maybe. 
See comments below.
For your search input, you are working a little bit tricky. I found a solution as following: https://github.com/JonasImm/pryces/blob/master/client/src/components/ProductSearchBar.js#L9
You could create a container with your search Icon, the input field, and your Cancel Button. This could be sorted with `display: flex` so you don't have to work with fixed pixel values. But just an idea :) If you like some input, let me know 💯 

Also, you could/should check your console before submitting your PR as there seem to be a few errors. (See below)
image

File: client/src/App.js (hunk 1 above)
Comment (rheimers) [reply client/src/App.js]: still have to warm up with rem, thx man 👍

File: client/src/components/SearchPageHeader.js (hunk 2 above)
Comment (rheimers) [reply client/src/components/SearchPageHeader.js]: Check

File: client/src/components/Searchbar.js (hunk 3 above)
Comment (rheimers) [reply client/src/components/Searchbar.js]: done

File: client/src/components/Searchbar.js (hunk 4 above)
Comment (rheimers) [reply client/src/components/Searchbar.js]: done

File: client/src/components/Searchbar.js (hunk 5 above)
Comment (rheimers) [reply client/src/components/Searchbar.js]: Thank you for the reminder

File: client/src/pages/SearchPage.js (hunk 12 above)
Comment (rheimers) [reply client/src/pages/SearchPage.js]: thx

File: client/src/pages/SearchPage.js (hunk 13 above)
Comment (rheimers) [reply client/src/pages/SearchPage.js]: Ah Jonas, no review without redundant console.logs

File: client/src/pages/SearchPage.js (hunk 16 above)
Comment (rheimers) [reply client/src/pages/SearchPage.js]: 💯

File: client/src/pages/SearchPage.js (hunk 15 above)
Comment (rheimers) [reply client/src/pages/SearchPage.js]: Yes, good point

File: client/src/pages/SearchPage.js (hunk 14 above)
Comment (rheimers) [reply client/src/pages/SearchPage.js:21]: I want to get/retrieve this param ("q") given in the url (http://localhost:3000/search?q=anna)

**Comment by rheimers** [comment]
Thanks a lot @JonasImm aka MASTER OF REVIEWS

> Hi 👍 Gj so far 💯
> I found a few things you should check maybe.
> See comments below.
> For your search input, you are working a little bit tricky. I found a solution as following: https://github.com/JonasImm/pryces/blob/master/client/src/components/ProductSearchBar.js#L9
> You could create a container with your search Icon, the input field, and your Cancel Button. This could be sorted with `display: flex` so you don't have to work with fixed pixel values. But just an idea :) If you like some input, let me know 💯
> 
> Also, you could/should check your console before submitting your PR as there seem to be a few errors. (See below)
> image

**Pull request merged**